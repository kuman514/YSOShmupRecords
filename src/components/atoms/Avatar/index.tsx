import styled from 'styled-components';

interface RootProps {
  $pxSize: number;
}

const Root = styled.div<RootProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: calc(${({ $pxSize }) => $pxSize}px / 2);
  width: ${({ $pxSize }) => $pxSize}px;
  height: ${({ $pxSize }) => $pxSize}px;
  background-color: var(--gray-color);
`;

const Image = styled.img`
  color: var(--black-color);

  font-size: 16px;
  font-weight: 500;

  border-radius: inherit;
  width: inherit;
  height: inherit;
`;

interface Props {
  pxSize: number;
  imageUrl: string;
}

export function Avatar({ pxSize, imageUrl }: Props) {
  return (
    <Root $pxSize={pxSize}>
      <Image src={imageUrl} alt="YSOShmupRecords avatar" />
    </Root>
  );
}
