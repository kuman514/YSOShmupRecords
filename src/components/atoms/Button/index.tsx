import React, { CSSProperties, ReactNode } from 'react';
import styled from 'styled-components';

import { ButtonType } from '^/types';

const backgroundColorByType: Record<ButtonType, string> = {
  [ButtonType.SOLID]: 'var(--primary-color)',
  [ButtonType.LINE]: 'transparent',
  [ButtonType.ROUND_SOLID]: 'var(--primary-color)',
  [ButtonType.ROUND_LINE]: 'transparent',
  [ButtonType.CLEAR]: 'transparent',
};

const colorByType: Record<ButtonType, string> = {
  [ButtonType.SOLID]: '#ffffff',
  [ButtonType.LINE]: 'var(--primary-color)',
  [ButtonType.ROUND_SOLID]: '#ffffff',
  [ButtonType.ROUND_LINE]: 'var(--primary-color)',
  [ButtonType.CLEAR]: '#ffffff',
};

const borderByType: Record<ButtonType, string> = {
  [ButtonType.SOLID]: '1px solid var(--primary-color)',
  [ButtonType.LINE]: '1px solid var(--primary-color)',
  [ButtonType.ROUND_SOLID]: '1px solid var(--primary-color)',
  [ButtonType.ROUND_LINE]: '1px solid var(--primary-color)',
  [ButtonType.CLEAR]: 'none',
};

const borderRadiusByType: Record<ButtonType, string> = {
  [ButtonType.SOLID]: '4px',
  [ButtonType.LINE]: '4px',
  [ButtonType.ROUND_SOLID]: '16px',
  [ButtonType.ROUND_LINE]: '16px',
  [ButtonType.CLEAR]: '16px',
};

const disabledColorByType: Record<ButtonType, string> = {
  [ButtonType.SOLID]: '#ffffff',
  [ButtonType.LINE]: '#ffffff',
  [ButtonType.ROUND_SOLID]: '#ffffff',
  [ButtonType.ROUND_LINE]: '#ffffff',
  [ButtonType.CLEAR]: 'var(--gray-color)',
};

const disabledBackgroundColorByType: Record<ButtonType, string> = {
  [ButtonType.SOLID]: 'var(--gray-color)',
  [ButtonType.LINE]: 'var(--gray-color)',
  [ButtonType.ROUND_SOLID]: 'var(--gray-color)',
  [ButtonType.ROUND_LINE]: 'var(--gray-color)',
  [ButtonType.CLEAR]: 'transparent',
};

const disabledTextDecorationLineByType: Record<ButtonType, string> = {
  [ButtonType.SOLID]: 'none',
  [ButtonType.LINE]: 'none',
  [ButtonType.ROUND_SOLID]: 'none',
  [ButtonType.ROUND_LINE]: 'none',
  [ButtonType.CLEAR]: 'line-through',
};

interface RootProps {
  $type: ButtonType;
}

const Root = styled.button<RootProps>`
  all: unset;
  box-sizing: border-box;

  font-family: Inter;
  font-size: 16px;
  font-weight: 500;

  padding: 8px 16px;

  color: ${({ $type }) => colorByType[$type]};
  background-color: ${({ $type }) => backgroundColorByType[$type]};
  border: ${({ $type }) => borderByType[$type]};
  border-radius: ${({ $type }) => borderRadiusByType[$type]};

  cursor: pointer;

  &:hover {
    background-color: var(--hovering-color);
    color: #ffffff;
  }

  &:disabled {
    border: none;
    color: ${({ $type }) => disabledColorByType[$type]};
    background-color: ${({ $type }) => disabledBackgroundColorByType[$type]};
    text-decoration: ${({ $type }) => disabledTextDecorationLineByType[$type]};
  }
`;

interface Props {
  type: ButtonType;
  children?: ReactNode;
  isDisabled: boolean;
  customStyle?: CSSProperties;
  onClick(): void;
}

export function Button({
  type,
  children,
  isDisabled,
  customStyle,
  onClick,
}: Props) {
  return (
    <Root
      $type={type}
      style={customStyle}
      disabled={isDisabled}
      onClick={onClick}
    >
      {children}
    </Root>
  );
}
