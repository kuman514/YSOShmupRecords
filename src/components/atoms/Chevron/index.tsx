import React from 'react';

import { ReactComponent as ClickToOpenChevronSvg } from '^/assets/chevrons/click-to-open-chevron.svg';
import { ReactComponent as ClickToCloseChevronSvg } from '^/assets/chevrons/click-to-close-chevron.svg';

interface Props {
  isOpen: boolean;
  fillColor: string;
}

export function Chevron({ isOpen, fillColor }: Props) {
  return isOpen ? (
    <ClickToCloseChevronSvg style={{ fill: fillColor }} />
  ) : (
    <ClickToOpenChevronSvg style={{ fill: fillColor }} />
  );
}
