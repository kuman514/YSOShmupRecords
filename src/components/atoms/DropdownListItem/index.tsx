import React from 'react';
import styled from 'styled-components';

const Root = styled.li`
  width: 240px;
`;

const Button = styled.button`
  border: unset;

  width: 100%;
  height: 100%;
  padding: 8px 15px;

  display: flex;
  align-items: center;

  color: var(--article-font-color);
  font-size: 12px;
  font-weight: 400;

  cursor: pointer;

  background: var(--active-color);

  &:hover {
    background: var(--hover-color);
  }

  &:disabled {
    background: var(--main-color);
  }
`;

interface Props {
  label: string;
  isSelected: boolean;
  onClick(): void;
}

export function DropdownListItem({ label, isSelected, onClick }: Props) {
  return (
    <Root>
      <Button disabled={isSelected} onClick={onClick}>
        {label}
      </Button>
    </Root>
  );
}
