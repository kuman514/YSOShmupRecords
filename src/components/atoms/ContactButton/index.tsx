import { styled } from 'styled-components';

import RawChatCircleDotsSvgrepoSvg from '^/assets/icons/chat-circle-dots-svgrepo-com.svg?react';
import { Button } from '^/components/atoms/Button';
import { ButtonType } from '^/types';

const ChatCircleDotsSvgrepoSvg = styled(RawChatCircleDotsSvgrepoSvg)`
  width: 36px;
  height: 36px;
  stroke: var(--white-color);
`;

interface Props {
  onClick(): void;
}

export function ContactButton({ onClick }: Props) {
  return (
    <Button
      type={ButtonType.ROUND_SOLID}
      customStyle={{
        position: 'fixed',
        bottom: '24px',
        right: '24px',
        zIndex: 21,
        width: '64px',
        height: '64px',
        padding: 0,
        borderRadius: '50%',
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
      }}
      isDisabled={false}
      onClick={onClick}
    >
      <ChatCircleDotsSvgrepoSvg />
    </Button>
  );
}
