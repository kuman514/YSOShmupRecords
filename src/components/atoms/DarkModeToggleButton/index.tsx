import { useEffect, useState } from 'react';
import { styled } from 'styled-components';

import RawMoonSvgrepoSvg from '^/assets/icons/moon-svgrepo-com.svg?react';
import RawSunSvgrepoSvg from '^/assets/icons/sun-svgrepo-com.svg?react';
import { Button } from '^/components/atoms/Button';
import {
  COLOR_THEME_LOCAL_STORAGE_KEY,
  isDarkModeTurnedOn,
} from '^/constants/dark-mode';
import { ButtonType } from '^/types';

const SunSvgrepoSvg = styled(RawSunSvgrepoSvg)`
  width: 36px;
  height: 36px;
  fill: var(--white-color);
`;

const MoonSvgrepoSvg = styled(RawMoonSvgrepoSvg)`
  width: 36px;
  height: 36px;
  fill: var(--white-color);
`;

export function DarkModeToggleButton() {
  const preferredTheme = window.matchMedia?.('(prefers-color-scheme: dark)')
    .matches
    ? 'dark'
    : 'light';
  const [colorTheme, setColorTheme] = useState<string>(
    localStorage.getItem(COLOR_THEME_LOCAL_STORAGE_KEY) ?? preferredTheme
  );

  function handleOnClickToggleColorTheme(newTheme: string) {
    localStorage.setItem(COLOR_THEME_LOCAL_STORAGE_KEY, newTheme);
    setColorTheme(newTheme);
  }

  const renderIcon = isDarkModeTurnedOn[colorTheme] ? (
    <MoonSvgrepoSvg />
  ) : (
    <SunSvgrepoSvg />
  );

  useEffect(() => {
    document.documentElement.setAttribute('color-theme', colorTheme);
  }, [colorTheme]);

  return (
    <Button
      type={ButtonType.CLEAR}
      customStyle={{
        padding: '4px',
        paddingBottom: 0,
        borderRadius: '50%',
      }}
      isDisabled={false}
      onClick={() => {
        switch (colorTheme) {
          case 'light':
            handleOnClickToggleColorTheme('dark');
            break;
          case 'dark':
            handleOnClickToggleColorTheme('light');
            break;
          default:
            break;
        }
      }}
    >
      {renderIcon}
    </Button>
  );
}
