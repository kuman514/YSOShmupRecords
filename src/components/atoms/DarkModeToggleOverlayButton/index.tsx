import React, { useEffect, useState } from 'react';

import { Button } from '^/components/atoms/Button';
import { ButtonType } from '^/types';
import { isDarkModeTurnedOn } from '^/constants/dark-mode';

export function DarkModeToggleOverlayButton() {
  const preferredTheme = window.matchMedia?.('(prefers-color-scheme: dark)')
    .matches
    ? 'dark'
    : 'light';
  const [colorTheme, setColorTheme] = useState<string>(
    localStorage.getItem('yso-shmup-records-color-theme') ?? preferredTheme
  );

  function handleOnClickToggleColorTheme(newTheme: string) {
    localStorage.setItem('yso-shmup-records-color-theme', newTheme);
    setColorTheme(newTheme);
  }

  const renderLabel = isDarkModeTurnedOn[colorTheme]
    ? '다크 모드 끄기'
    : '다크 모드 켜기';

  useEffect(() => {
    document.documentElement.setAttribute('color-theme', colorTheme);
  }, [colorTheme]);

  return (
    <Button
      type={ButtonType.ROUND_SOLID}
      customStyle={{
        position: 'fixed',
        bottom: '32px',
        right: '32px',
        zIndex: 10,
      }}
      isDisabled={false}
      onClick={() => {
        switch (colorTheme) {
          case 'light':
            handleOnClickToggleColorTheme('dark');
            break;
          case 'dark':
            handleOnClickToggleColorTheme('light');
            break;
          default:
            break;
        }
      }}
    >
      {renderLabel}
    </Button>
  );
}
