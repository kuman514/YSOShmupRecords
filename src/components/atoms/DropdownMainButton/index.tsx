import React from 'react';
import styled from 'styled-components';

import { Chevron } from '^/components/atoms/Chevron';

const Root = styled.button`
  border: unset;

  width: 240px;
  padding: 8px 15px;

  display: flex;
  justify-content: space-between;
  align-items: center;

  cursor: pointer;

  background: var(--active-color);
`;

const Label = styled.span`
  color: var(--article-font-color);
  font-size: 12px;
  font-weight: 400;
`;

interface Props {
  label: string;
  isOpen: boolean;
  onClick(): void;
}

export function DropdownMainButton({ label, isOpen, onClick }: Props) {
  return (
    <Root onClick={onClick}>
      <Label>{label}</Label>
      <Chevron isOpen={isOpen} fillColor="var(--article-font-color)" />
    </Root>
  );
}
