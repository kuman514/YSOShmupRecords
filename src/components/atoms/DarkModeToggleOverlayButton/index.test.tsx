import React from 'react';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import 'jest-styled-components';

import { DarkModeToggleOverlayButton } from '^/components/atoms/DarkModeToggleOverlayButton';

describe('DarkModeToggleOverlayButton', () => {
  beforeEach(() => {
    cleanup();
  });

  it('should have a snapshot match', () => {
    const { container } = render(
      <DarkModeToggleOverlayButton onClick={() => {}} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should have a clickable button and an unclickable overlay', async () => {
    const mockFn = vi.fn();
    render(<DarkModeToggleOverlayButton onClick={mockFn} />);

    fireEvent.click(await screen.findByTestId('toggle-button'));
    expect(mockFn).toBeCalledTimes(1);
  });
});
