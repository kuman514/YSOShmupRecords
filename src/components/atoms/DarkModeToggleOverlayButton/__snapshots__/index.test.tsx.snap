// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DarkModeToggleOverlayButton > should have a snapshot match 1`] = `
.c0 {
  all: unset;
  box-sizing: border-box;
  font-size: 16px;
  font-weight: 500;
  padding: 8px 16px;
  color: var(--white-color);
  background-color: var(--primary-color);
  border: 1px solid var(--primary-color);
  border-radius: 16px;
  cursor: pointer;
}

.c0:hover {
  background-color: var(--hovering-color);
  color: var(--white-color);
}

.c0:disabled {
  border: none;
  color: var(--white-color);
  background-color: var(--gray-color);
  text-decoration: none;
}

<button
  class="c0"
  style="position: fixed; bottom: 32px; right: 32px;"
>
  다크 모드 켜기
</button>
`;
